app/models/server/models/Events.ts

```diff
@@ -0,0 +1,161 @@
+		}
+
+		const stub: IEventStub<IEDataGenesis> = {
+			t: EventTypeDescriptor.GENESIS,
+			d,
+		};
+
+		return this.createEvent(src, contextQuery, stub);
+	}
+
+	public async addEvent<T extends EDataDefinition>(contextQuery: IContextQuery, event: IEvent<T>): Promise<IAddEventResult> {
+		// Check if the event does not exit
+		const existingEvent = this.findOne({ _id: event._id });
+
+		// If it does not, we insert it, checking for the parents
+		if (!existingEvent) {
+			// Check if we have the parents
+			const parents: Array<IEvent<any>> = await this.model.rawCollection().find({ ...contextQuery, _id: { $in: event._pids } }, { _id: 1 }).toArray();
+			const _pids: Array<string> = parents.map((e: IEvent<any>) => e._id);
+
+			// This means that we do not have the parents of the event we are adding
+			if (_pids.length !== event._pids.length) {
+				const { src } = event;
+
+				// Get the latest events for that context and src
+				const latestEvents = await this.model.rawCollection().find({ ...contextQuery, src }, { _id: 1 }).toArray();
+				const latestEventIds = latestEvents.map((e: IEvent<any>) => e._id);
+
+				return {
+					success: false,
+					reason: 'missingParents',
+					missingParentIds: event._pids.filter((_id) => _pids.indexOf(_id) === -1),
+					latestEventIds,
+				};
+			}
+
+			// Clear the "hasChildren" of the parent events
+			await this.update({ _id: { $in: _pids } }, { $unset: { hasChildren: '' } }, { multi: 1 });
+
+			this.insert(event);
+		}
+
+		return {
+			success: true,
+		};
+	}
+
+	private async getExistingEvent(contextQuery: IContextQuery, eventCID: string, eventT: string): Promise<IEvent<any>> {
+		return this.model
+			.rawCollection()
+			.findOne({ ...contextQuery, _cid: eventCID, t: eventT.substr(1) });
+	}
+
+	public async updateEventData<T extends EDataDefinition>(contextQuery: IContextQuery, eventCID: string, eventT: string, updateData: IEDataUpdate<T>): Promise<void> {
+		const existingEvent = await this.getExistingEvent(contextQuery, eventCID, eventT);
+
+		const updateQuery: any = EJSON.fromJSONValue(deepMapKeys(updateData, (k: any) => k.replace('_csg', '$').replace('_dot', '.')));
+
+		for (const prop in updateQuery) {
+			if (prop.startsWith('$')) {
+				updateQuery[prop] = _.mapKeys(updateQuery[prop], (value: any, key: any) => (key.startsWith('$') ? key : `d.${ key }`));
+			}
+		}
+
+		await this.model.rawCollection().update({ _id: existingEvent._id }, updateQuery);
+	}
+
+	public async flagEventAsDeleted(contextQuery: IContextQuery, eventCID: string, eventT: string, deletedAt: Date): Promise<void> {
+		const existingEvent = await this.getExistingEvent(contextQuery, eventCID, eventT);
+
+		await this.model.rawCollection().update({ _id: existingEvent._id }, {
+			$set: {
+				_deletedAt: deletedAt,
+			},
+		});
+	}
+
+	// async getEventById(contextQuery: ContextQuery, eventId:string) {
```

Comment:
Can we remove these lines?

Reply:
I will review this, but it is probably legacy, related to federation